### PR TITLE
 PERF/ENH: add fast astyping for Categorical

### LINKED ISSUE #8628: ENH: decode for Categoricals
There seems no direct way to return to the original dtype and the [documentation](http://pandas-docs.github.io/pandas-docs-travis/categorical.html#object-creation) recommends: _"To get back to the original Series or numpy array, use Series.astype(original_dtype) or np.asarray(categorical)"_

That's slow and a `decode` or `decat` method would be trivial:

``` python
df=pd.DataFrame(np.random.choice(list(u'abcde'), 4e6).reshape(1e6, 4),
    columns=list(u'ABCD'))                                     
for col in df.columns: df[col] = df[col].astype('category')   

%timeit for col in df.columns: df[col].astype('unicode')      
1 loops, best of 3: 1.06 s per loop

%timeit for col in df.columns: cats=df[col].cat.categories; cats[df[col].cat.codes]    
10 loops, best of 3: 33.2 ms per loop   
```

I was working with ~10 categories (partially longer strings) on a 20 mio rows dataset where the difference was even bigger (unfortunately can't reproduce it with dummy data) and using `astype` felt rather buggy (minutes) than only a performance issue.

Given the current limitations on exporting categorical data, having a fast `decode` method would be very convenient. Since category codes are most often strings an optional parameter for direct character set encoding would also be good to have for such a method.

``` python
%timeit for col in df.columns: df[col].astype('unicode').str.encode('latin1')  
1 loops, best of 3: 3.95 s per loop
%timeit for col in df.columns: cats=pd.Series(df[col].cat.categories).str.encode('latin1'); cats[df[col].cat.codes]                                                                  
10 loops, best of 3: 74.5 ms per loop   
```

### PR DESCRIPTION
**Pull request by arw2019**
- [x] closes #8628
- [ ] tests added / passed
- [x] benchmarks added
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

To illustrate the speed-up, the set-up is (from OP):
```python
import numpy as np
import pandas as pd

rng = np.random.default_rng()

df = pd.DataFrame(
    rng.choice(np.array(list("abcde")), 4_000_000).reshape(1_000_000, 4),
    columns=list("ABCD"),
)

for col in df.columns:
    df[col] = df[col].astype("category")
```
On master
``` python
In [5]: %timeit [df[col].astype('unicode') for col in df.columns] 
250 ms ± 601 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
versus on this branch
``` python
In [5]: %timeit [df[col].astype('unicode') for col in df.columns] 
5.38 ms ± 47 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

### CODE REVIEW

File: pandas/core/generic.py
Comment (jbrockmendel) [inline pandas/core/generic.py:5799]: this probably belongs on `core.internals.blocks.Block.astype`; `is_categorical_dtype(self)` doesnt make sense for DataFrame (also, for perf, always better to pass a dtype object to is_foo_dtype rather than an arraylike object)

Comment (arw2019) [reply pandas/core/generic.py]: Right. Done (& noted re: dtype checks)

**Comment by topper-123** [comment]
This changes the behaviour unfortunately:

```python
>>> n = 100_000
>>> cat = pd.Categorical(["a", "b"] * n)
>>> ser = pd.Series(cat)
>>> ser.astype(str).dtype
dtype('O')  # master
CategoricalDtype(categories=['a', 'b'], ordered=False)  # this PR, different dtype
```
Can you make it work in some different way?

**Comment by arw2019** [comment]
> This changes the behaviour unfortunately:
> 
> ```python
> >>> n = 100_000
> >>> cat = pd.Categorical(["a", "b"] * n)
> >>> ser = pd.Series(cat)
> >>> ser.astype(str).dtype
> dtype('O')  # master
> CategoricalDtype(categories=['a', 'b'], ordered=False)  # this PR, different dtype
> ```
> 
> Can you make it work in some different way?

Thanks for the catch! Fixed this now

File: pandas/core/internals/blocks.py
Comment (arw2019) [inline pandas/core/internals/blocks.py:602]: Could define a new method for this in `core/dtypes/common.py` (`is_string_like_or_object_dtype`) ?

**Comment by arw2019** [comment]
It seems like adding this to astype has a lot of repercussions downstream, casting objects in undesirable ways. The special casing needed to make the tests pass is a code smell. 

I wonder if it might make sense, instead of adding this to `astype`,  to create a dedicated method for astyping a `Categorical` via this method?
```
In []: cat = pd.Categorical(["a", "b", "a", "a"] ) 
   ...: s = pd.Series(cat) 
   ...: s.astype_categories(dtype=str)                                      
Out[]: 
0    a
1    b
2    a
3    a
dtype: object
```

**Comment by jreback** [comment]
see also #37371; this might be the same path.

File: pandas/core/internals/blocks.py
Comment (jbrockmendel) [inline pandas/core/internals/blocks.py:610]: this seems really convoluted, in a method that is already too complicated as it is (xref #22369)

Do you have a good idea where the perf improvement comes from?  e.g. could we push this down into `Categorical.astype`?

Comment (arw2019) [reply pandas/core/internals/blocks.py:610]: Agreed that the amount of special casing here is not good (and even with that a bunch of tests are still failing)

The perf improvement is from astyping just the category labels instead of astyping each array entry separately.

`Categorical.astype` seems like a good location for this - will give that a go

File: pandas/core/arrays/categorical.py
Comment (jreback) [inline pandas/core/arrays/categorical.py:419]: you don't need to pass dtype here?

File: pandas/core/arrays/categorical.py
Comment (jreback) [inline pandas/core/arrays/categorical.py:417]: add a comment here on why we are doing this

File: asv_bench/benchmarks/categoricals.py
Comment (jreback) [inline asv_bench/benchmarks/categoricals.py:84]: can you add benchmarks for other types of categories (int, dti) for example. show the results of the benchmarks.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/categorical.py
Comment (arw2019) [reply pandas/core/arrays/categorical.py:421]: I think no because we've already astyped the category a few lines up

**Comment by arw2019** [comment]
Re: review comments:

#### `int` benchmark results
```
In [1]: import numpy as np 
   ...: import pandas as pd 
   ...:  
   ...: rng = np.random.default_rng() 
   ...:  
   ...: df = pd.DataFrame( 
   ...:     rng.choice(np.arange(8), 4_000_000).reshape(1_000_000, 4), 
   ...:     columns=list("ABCD"), 
   ...: ) 
   ...:  
   ...: for col in df.columns: 
   ...:     df[col] = df[col].astype("category") 
   ...:                                                                                                 
```
On master:
```
In [2]: %timeit [df[col].astype('int') for col in df.columns]                                           
20.9 ms ± 181 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```
On this branch:
```
In [4]: %timeit [df[col].astype('int') for col in df.columns]                                                                                                                                                      
18.7 ms ± 115 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

File: pandas/core/arrays/categorical.py
Comment (arw2019) [reply pandas/core/arrays/categorical.py:419]: Done

File: asv_bench/benchmarks/categoricals.py
Comment (arw2019) [reply asv_bench/benchmarks/categoricals.py:84]: Ok!

Posted int benchmark in main thread + will add/post more

**Comment by jreback** [comment]
this is not showing much of an improvement. can you try someother dtypes as well.

**Comment by arw2019** [comment]
> this is not showing much of an improvement. can you try someother dtypes as well.

There's a decent (~40x) speed up with string->float:
``` python
In [3]: import numpy as np 
   ...: import pandas as pd 
   ...:  
   ...: rng = np.random.default_rng() 
   ...:  
   ...: cats = [str(np.random.random_sample()) for _ in range(10)] 
   ...: df = pd.DataFrame( 
   ...:     rng.choice(np.array(cats), 4_000_000).reshape(1_000_000, 4), 
   ...:     columns=list("ABCD"), 
   ...: ) 
   ...:  
   ...: for col in df.columns: 
   ...:     df[col] = df[col].astype("category") 
   ...:                                                                                                   
```
On master
```
In [4]: %timeit [df[col].astype('float') for col in df.columns]                 
873 ms ± 3.69 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
versus on this branch:
```
In [4]: %timeit [df[col].astype('float') for col in df.columns]                                           
23.2 ms ± 1.09 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

File: asv_bench/benchmarks/categoricals.py
Comment (jreback) [reply asv_bench/benchmarks/categoricals.py:84]: right pls update these for int,float,string,datetime

File: asv_bench/benchmarks/categoricals.py
Comment (jreback) [inline asv_bench/benchmarks/categoricals.py:74]: 10**5 is prob enough here

**Comment by jreback** [comment]
> > this is not showing much of an improvement. can you try someother dtypes as well.
> 
> There's a decent (~40x) speed up with string->float:
> 
> ```python
> In [3]: import numpy as np 
>    ...: import pandas as pd 
>    ...:  
>    ...: rng = np.random.default_rng() 
>    ...:  
>    ...: cats = [str(np.random.random_sample()) for _ in range(10)] 
>    ...: df = pd.DataFrame( 
>    ...:     rng.choice(np.array(cats), 4_000_000).reshape(1_000_000, 4), 
>    ...:     columns=list("ABCD"), 
>    ...: ) 
>    ...:  
>    ...: for col in df.columns: 
>    ...:     df[col] = df[col].astype("category") 
>    ...:                                                                                                   
> ```
> 
> On master
> 
> ```
> In [4]: %timeit [df[col].astype('float') for col in df.columns]                 
> 873 ms ± 3.69 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
> ```
> 
> versus on this branch:
> 
> ```
> In [4]: %timeit [df[col].astype('float') for col in df.columns]                                           
> 23.2 ms ± 1.09 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
> ```

right pls show for all dtypes

File: doc/source/whatsnew/v1.2.0.rst
Comment (jreback) [inline doc/source/whatsnew/v1.2.0.rst:350]: its actually for Series.astype, but you can mention for both.

**Review by jreback — CHANGES_REQUESTED** [review]
looks good, pls merge master and address comments, ping on green.

File: asv_bench/benchmarks/categoricals.py
Comment (arw2019) [reply asv_bench/benchmarks/categoricals.py:87]: Ok

File: doc/source/whatsnew/v1.2.0.rst
Comment (arw2019) [reply doc/source/whatsnew/v1.2.0.rst]: Done

**Comment by arw2019** [comment]
Addressed comments, waiting on CI (I suspect some test failures are real so may need to address that)

**Comment by jreback** [comment]
lgtm. merge master and ping on green.

**Comment by arw2019** [comment]
@jreback CI greenish but this probably needs another review because I made a few changes since you last looked (caught edge cases, cleaned up the if/elif/else and consolidated the return)

File: pandas/core/arrays/categorical.py
Comment (jreback) [inline pandas/core/arrays/categorical.py:431]: we *never* should add the na_value into the categories themselves. this should have broken things. (codes indicates a -1) for the missing value.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/categorical.py
Comment (jreback) [inline pandas/core/arrays/categorical.py:431]: hmm this is not exactly an accurate message. I think you need to catch the message and re-raise something better. (e.g. Index is very misleading here) - i am referring to the changes you made in the tests.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/categorical.py
Comment (arw2019) [reply pandas/core/arrays/categorical.py:431]: I thought this isn't the right way to do it. Though this actually this fixes the tests because in the slice `new_categories[self.codes]` we need `-1` to pick out the na_value and with this it does

(This is why the previous variant I had wasn't going green)

Comment (arw2019) [reply pandas/core/arrays/categorical.py]: > we _never_ should add the na_value into the categories themselves. this should have broken things. (codes indicates a -1) for the missing value.

Rewrote this using `take_1d` like we do in some other methods. Passes all tests modulo the error messages (commented below)

Comment (arw2019) [reply pandas/core/arrays/categorical.py]: I'm catching this in the new commit

File: pandas/core/arrays/categorical.py
Comment (jbrockmendel) [inline pandas/core/arrays/categorical.py:427]: i think do `self.categories.astype` and then cast to numpy if necessary afterwards, may avoid an object-dtype cast in the middle

Comment (arw2019) [reply pandas/core/arrays/categorical.py:427]: Yeah that cuts the time. Before:
```
In [2]: %timeit [df[col].astype('unicode') for col in df.columns]
106 ms ± 28.1 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```
and after
```
In [2]: %timeit [df[col].astype('unicode') for col in df.columns]
52.8 ms ± 4.77 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

File: pandas/core/arrays/categorical.py
Comment (jreback) [inline pandas/core/arrays/categorical.py:432]: pretty sure you are copying twice here as the take_1d copies.

Comment (jreback) [reply pandas/core/arrays/categorical.py:432]: of if we could have an EA returned, then do `np.asarray`

Comment (jreback) [inline pandas/core/arrays/categorical.py:432]: you need to do `ensure_platform_int(self._code)` here to avoid 32-bit failing.

**Review by jreback — CHANGES_REQUESTED** [review]
looks good, the 32-bit

File: pandas/core/arrays/categorical.py
Comment (arw2019) [reply pandas/core/arrays/categorical.py]: EA case is handled above so I removed here

But probably it's possible to consolidate EA/non-EA and get the same speed up for EAs (will look in follow-on)

Comment (arw2019) [reply pandas/core/arrays/categorical.py]: Thanks!!! Added

Comment (arw2019) [reply pandas/core/arrays/categorical.py]: > But probably it's possible to consolidate EA/non-EA and get the same speed up for EAs (will look in follow-on)

Added a TODO for this

Comment (arw2019) [reply pandas/core/arrays/categorical.py]: Hmmm I think I didn't do this right - 32 bit build still failing
```
(...)
        result = cat.astype(int)
        expected = np.array(cat, dtype=int)
>       tm.assert_numpy_array_equal(result, expected)
E       AssertionError: numpy array are different
E       
E       Attribute "dtype" are different
E       [left]:  int64
E       [right]: int32

```

Comment (jbrockmendel) [reply pandas/core/arrays/categorical.py]: maybe pass `dtype=self._codes.dtype` to take_1d?

Comment (arw2019) [reply pandas/core/arrays/categorical.py]: Tried that, take_1d doesn't take a dtype arg...

Comment (jbrockmendel) [reply pandas/core/arrays/categorical.py]: sorry, i meant to np.array

Comment (arw2019) [reply pandas/core/arrays/categorical.py]: Done (CI green now)

File: pandas/core/arrays/categorical.py
Comment (jbrockmendel) [inline pandas/core/arrays/categorical.py:431]: why change TypeError?

**Comment by jreback** [comment]
ci / checks failing

File: pandas/core/arrays/categorical.py
Comment (arw2019) [reply pandas/core/arrays/categorical.py:431]: It's to fix the error message for `CategoricalIndex`. If we don't catch `TypeError` we end up with `TypeError: Cannot cast Index to dtype float64` (below) versus something like `TypeError: Cannot cast object to dtype float64`
``` python
In [2]: idx = pd.CategoricalIndex(["a", "b", "c", "a", "b", "c"])

In [3]: idx.astype('float')
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
/workspaces/pandas-arw2019/pandas/core/indexes/base.py in astype(self, dtype, copy)
    700         try:
--> 701             casted = self._values.astype(dtype, copy=copy)
    702         except (TypeError, ValueError) as err:

ValueError: could not convert string to float: 'a'

The above exception was the direct cause of the following exception:

TypeError                                 Traceback (most recent call last)
<ipython-input-4-38d56ec15c36> in <module>
----> 1 idx.astype('float')

/workspaces/pandas-arw2019/pandas/core/indexes/category.py in astype(self, dtype, copy)
    369     @doc(Index.astype)
    370     def astype(self, dtype, copy=True):
--> 371         res_data = self._data.astype(dtype, copy=copy)
    372         return Index(res_data, name=self.name)
    373 

/workspaces/pandas-arw2019/pandas/core/arrays/categorical.py in astype(self, dtype, copy)
    427             # GH8628 (PERF): astype category codes instead of astyping array
    428             try:
--> 429                 astyped_cats = self.categories.astype(dtype=dtype, copy=copy)
    430             except (ValueError):
    431                 raise ValueError(

/workspaces/pandas-arw2019/pandas/core/indexes/base.py in astype(self, dtype, copy)
    701             casted = self._values.astype(dtype, copy=copy)
    702         except (TypeError, ValueError) as err:
--> 703             raise TypeError(
    704                 f"Cannot cast {type(self).__name__} to dtype {dtype}"
    705             ) from err

TypeError: Cannot cast Index to dtype float64
```

**Comment by arw2019** [comment]
> ci / checks failing

Hmmm the mypy complaint looks unrelated (also getting the same thing on my other PRs)

```
mypy --version
mypy 0.782
Performing static analysis using mypy
pandas/core/indexes/datetimelike.py:776: error: Argument 1 to "_simple_new" of "DatetimeIndexOpsMixin" has incompatible type "Union[ExtensionArray, Any]"; expected "Union[DatetimeArray, TimedeltaArray, PeriodArray]"  [arg-type]
Found 1 error in 1 file (checked 1119 source files)
Performing static analysis using mypy DONE
```
Anyway will keep looking and ping when green

**Comment by jreback** [comment]
> > ci / checks failing
> 
> Hmmm the mypy complaint looks unrelated (also getting the same thing on my other PRs)
> 
> ```
> mypy --version
> mypy 0.782
> Performing static analysis using mypy
> pandas/core/indexes/datetimelike.py:776: error: Argument 1 to "_simple_new" of "DatetimeIndexOpsMixin" has incompatible type "Union[ExtensionArray, Any]"; expected "Union[DatetimeArray, TimedeltaArray, PeriodArray]"  [arg-type]
> Found 1 error in 1 file (checked 1119 source files)
> Performing static analysis using mypy DONE
> ```
> 
> Anyway will keep looking and ping when green

oh i think that's fixed on master

File: pandas/core/arrays/categorical.py
Comment (jreback) [reply pandas/core/arrays/categorical.py:431]: ok this is fine, but can you add a comment for this then right here (so future readers understand)

**Review by jreback — CHANGES_REQUESTED** [review]
small comment, ping on green.

**Comment by arw2019** [comment]
@jreback Green + addressed comment

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @arw2019 very nice

**Comment by arw2019** [comment]
thanks @jreback @jbrockmendel for the reviews!

**Pull request merged**
